File: nn/Dense.cpp
#include "Dense.h"


evo::nn:: Dense::Dense(int unit_size, std::string activation, std::vector<float> weight_params, std::vector<float> bias_params,
	bool bias_bool, bool store_vectors, std::string fp, float alpha)
	:out_size(unit_size), in_size(0), activation(activation), weight_params(weight_params), bias_params(bias_params), bias_bool(bias_bool),
	store_vectors(store_vectors), timestamp(1), fp(fp)
{
	act = evo::nn::activation(alpha);
	if (fp == "") {
		initialize_layer();
	}
}

evo::nn::Dense::Dense(int unit_size, int input_size, std::string activation, std::vector<float> weight_params, std::vector<float> bias_params,
	bool bias_bool, bool store_vectors, std::string fp, float alpha)
	:out_size(unit_size), in_size(input_size), activation(activation), weight_params(weight_params), bias_params(bias_params), bias_bool(bias_bool),
	store_vectors(store_vectors), timestamp(1), fp(fp)
{
	act = evo::nn::activation(alpha);
	if (fp == "") {
		initialize_layer();
	}
}

std::vector<float> evo::nn:: Dense::feedforward(std::vector<float> x) {
	vectors.insert(std::pair<std::string, std::vector<float>> ("x" + std::to_string(timestamp), x));
	std::vector<float> h = evo::matmul(x, weight);
	vectors.insert(std::pair<std::string, std::vector<float>>("h" + std::to_string(timestamp), h));
	std::vector<float> y = act.activate(h, activation, "a"+std::to_string(timestamp));
	vectors.insert(std::pair<std::string, std::vector<float>>("y" + std::to_string(timestamp), y));
	timestamp++;
	return y;
}

void evo::nn::Dense::train(std::vector<float> loss_vector, float training_rate) {
	for (int t = timestamp - 1; t > 0; t--) {

		for (int j = 0; j < out_size; j++) {
			float l = loss_vector[j];

			for (int i = 0; i < in_size; i++) {
				gradient(l, i, j, t, training_rate);
			}
		}
	}
}

void evo::nn::Dense::gradient(float error, int i, int j, int timestamp, float training_rate) {
	float dy_dh = act.get_error("a" + std::to_string(timestamp))[j];
	std::string vc = "x" + std::to_string(timestamp);
	std::vector<float> x = vectors[vc];
	float dh_w = 0;
	dh_w = x[i];
	float dh_dx = weight[i][j];
	float cost_w = error * dy_dh * training_rate, cost_b = error* dy_dh* training_rate;
	cost_w *= dh_w;
	weight[i][j] += cost_w;
	bias[j] += cost_b;
}

void evo::nn::Dense::initialize_layer() {
	weight = evo::random_mtx(in_size, out_size, weight_params[0], weight_params[1]);
	bias = evo::random_vec(out_size, weight_params[0], weight_params[1]);
}

template<typename T>
void evo::nn::Dense::print_vector(std::vector<T> a) {
	std::cout << "{ ";
	for (T e : a) {
		std::cout << e << " ";
	}
	std::cout << " }" << std::endl;
}

File: nn/Dense.py
from nn.Variables import Variables
from nn.Activation import Activation
# from Optimizer.Losses import Losses
import numpy as np


class Dense(Variables):

    def __init__(self, unit_size, input_size=None, activation='softmax', weight_params=(-1, 1), bias_params=(0, 1),
                 bias_bool=True, store_vectors=True, ):
        self.store_vectors = store_vectors
        super().__init__(unit_size, input_size, activation, weight_params, bias_params, bias_bool)

        self.A = Activation()
        self.Vectors = {}
        if not bias_bool:
            self.bias_params = (0, 0)

    def feed_forward(self, input_vect):
        x = np.asarray(input_vect)
        h = np.add(np.matmul(x, self.Weight), self.Bias)
        y = self.A.activation_function(h, self.activation)
        if self.store_vectors:
            self.Vectors['x'], self.Vectors['h'], self.Vectors['y'] = x, h, y
        return y

    def gradient(self, error, i, j, training_rate=0.6, momentum_bool=0, gamma=0.9, last_layer=True):
        assert isinstance(error, float) or isinstance(error, int)
        dy_dh = self.A.errors['a'][j]
        dh_w = self.Vectors['x'][i]
        # dh_dx = self.Weight[i][j]
        cost_w, cost_b = error * dy_dh * training_rate, error * dy_dh * training_rate
        if not last_layer:
            cost_w *= dh_w
        delta = momentum_bool*self.W_p_mtx[i][j] + cost_w
        self.Weight[i][j] += delta
        self.errors[i][j] += delta
        self.Bias[j] += momentum_bool*self.B_p_mtx[j] + cost_b
        if momentum_bool:
            self.momentum_handler(i, j, cost_w, cost_b, gamma)
        # return error*dy_dh*dh_dx

    def train(self, loss_vect, training_rate=0.5):
        for j in range(self.unit_size):
            l = loss_vect[j]
            for i in range(self.input_size):
                self.gradient(l, i, j, training_rate)

    def momentum_handler(self, i, j, cost_w, cost_b, gamma):
        self.W_p_mtx[i, j] = gamma*self.W_p_mtx[i, j] + cost_w
        self.B_p_mtx[j] = gamma*self.B_p_mtx[j] + cost_b
        return

    def transformLossTensor(self, loss_tensor):
        assert np.ndim(loss_tensor) == 1
        w = np.linalg.pinv(self.Weight)
        return np.matmul(loss_tensor, w)


def word2vectGenerator(vocab_size):
    lst = np.zeros(vocab_size)
    lst[np.random.randint(0, vocab_size)] = 1
    return lst


def labelGenerator(vocab_size):
    indices = np.random.randint(0, vocab_size, np.random.randint(1, vocab_size-1))
    lst = np.zeros(vocab_size)
    for index in indices:
        lst[index] = 1
    return lst


if __name__ == "__main__":
    i, o = 10, 4
    D = Dense(o, i, activation="sigmoid")
    x = word2vectGenerator(i)
    y = labelGenerator(o)

    for _ in range(10):
        z = D.feed_forward(x)
        l = np.subtract(y, z)
        D.train(l)
        print(z)
    print(y)



